src/drivers/fluid_wasapi.c

```diff
@@ -0,0 +1,811 @@
+ *    Vista doesn't seem to support the resampling method with better quality.
+ *  - Under Windows 10, this driver may report a latency of 0ms in shared mode.
+ *    This is nonsensical and should be disregarded.
+ * 
+ */
+
+#define FLUID_WASAPI_MAX_OUTPUTS 1
+
+typedef void(*fluid_wasapi_devenum_callback_t)(IMMDevice *, void *);
+
+static DWORD WINAPI fluid_wasapi_audio_run(void* p);
+static int fluid_wasapi_write_processed_channels(void *data, int len,
+                               int channels_count,
+                               void *channels_out[], int channels_off[],
+                               int channels_incr[]);
+static void fluid_wasapi_foreach_device(fluid_wasapi_devenum_callback_t callback, void *data);
+static void fluid_wasapi_register_callback(IMMDevice *dev, void *data);
+static void fluid_wasapi_finddev_callback(IMMDevice *dev, void *data);
+static IMMDevice* fluid_wasapi_find_device(IMMDeviceEnumerator *denum, const char *name);
+static FLUID_INLINE int16_t round_clip_to_i16(float x);
+
+typedef struct
+{
+    const char *name;
+    wchar_t *id;
+} fluid_wasapi_finddev_data_t;
+
+typedef struct
+{
+    fluid_audio_driver_t driver;
+
+    void *user_pointer;
+    fluid_audio_func_t func;
+    fluid_audio_channels_callback_t write;
+    float **drybuf;
+
+    UINT32 nframes;
+    double buffer_duration;
+    int channels_count;
+    int float_samples;
+
+    HANDLE thread;
+    DWORD thread_id;
+    HANDLE quit_ev;
+    
+    IAudioClient *aucl;
+    IAudioRenderClient *arcl;
+
+} fluid_wasapi_audio_driver_t;
+
+fluid_audio_driver_t *new_fluid_wasapi_audio_driver(fluid_settings_t *settings, fluid_synth_t *synth)
+{
+    return new_fluid_wasapi_audio_driver2(settings, (fluid_audio_func_t)fluid_synth_process, synth);
+}
+
+fluid_audio_driver_t *new_fluid_wasapi_audio_driver2(fluid_settings_t *settings, fluid_audio_func_t func, void *data)
+{
+    fluid_wasapi_audio_driver_t *dev = NULL;
+    double sample_rate;
+    int periods, period_size;
+    long long buffer_duration_reftime;
+    long long periods_reftime;
+    long long latency_reftime;
+    int audio_channels;
+    int sample_size;
+    int i;
+    char *dname = NULL;
+    DWORD flags = 0;
+    GUID guid_float = {DEFINE_WAVEFORMATEX_GUID(WAVE_FORMAT_IEEE_FLOAT)};
+    WAVEFORMATEXTENSIBLE wfx;
+    WAVEFORMATEXTENSIBLE *rwfx = NULL;
+    AUDCLNT_SHAREMODE share_mode;
+    IMMDeviceEnumerator *denum = NULL;
+    IMMDevice *mmdev = NULL;
+    HRESULT ret;
+    OSVERSIONINFOEXW vi = {sizeof(vi), 6, 0, 0, 0, {0}, 0};
+
+    if(!VerifyVersionInfoW(&vi, VER_MAJORVERSION|VER_MINORVERSION|VER_SERVICEPACKMAJOR,
+        VerSetConditionMask(VerSetConditionMask(VerSetConditionMask(0,
+            VER_MAJORVERSION,VER_GREATER_EQUAL),
+            VER_MINORVERSION,VER_GREATER_EQUAL),
+            VER_SERVICEPACKMAJOR, VER_GREATER_EQUAL)))
+    {
+        FLUID_LOG(FLUID_ERR, "wasapi: this driver requires Windows Vista or newer.");
```

Comment:
Do we need to have this "runtime" check? I would have expected cmake to complain (or silently not build wasapi) if you'd try to build it on e.g. XP.

Reply:
WASAPI uses COM, and therefore only needs the headers to build. Mingw always ship those headers, so I'd imagine it's possible to build this driver on Windows XP. If this driver is used on Windows XP or lower, `CoCreateInstance` would return `REGDB_E_CLASSNOTREG` and could be sort of confusing for end users.

Reply:
Ok.